design/transaction-layer-refactoring.md

```diff
@@ -0,0 +1,151 @@
+
+And now, different commands have different dependencies. But because all commands share the same procedure, we must pass in the union of all dependencies of all commands. It makes adding commands and dependencies more difficult.
+
+## New structure
+
+Based on the analysis above, the current structure is not flexible enough and should be dropped. Instead, we can change to put logic of a single command together. Common steps can be extracted as methods for reuse. Then, it will be much easier to find out the procedure of each command while still not repeating code.
+
+Steps like acquiring latches and waiting for raft replication cannot finish immediately, so it is appropriate to make every command an `async fn`.
+
+For example, the `AcquirePessimisticLock` command can be like:
+
+```rust
+async fn acquire_pessimistic_lock(&self, req: PessimisticLockRequest) -> PessimisticLockResponse {
+    let mut resp = PessimisticLockResponse::default();
+    let guards = self.lock_keys(req.mutations.iter().map(|m| m.get_key())).await;
+    let snapshot = self.get_snapshot(req.get_context()).await;
+    
+    let mut txn = MvccTxn::new(...);
+    let mut locked = None;
+    for (mutation, guard) in req.mutations.into_iter().zip(&guards) {
+        match txn.acquire_pessimistic_lock(...) {
+            Ok(...) => ...,
+            Err(KeyIsLocked(lock_info)) => {
+                guard.set_lock(lock_info.clone());
+                let lock_released = guard.lock_released(); // returns a future which is ready when the lock is released
+                locked = Some(lock_info, );
+            },
+            Err(e) => ...
+        }
+    }
+    
+    if let Some((lock_info, lock_released)) = locked {
+        drop(guards); // release the latches first
+        lock_released.await; // easy to add timeout
+        resp.set_errors(...);
+    } else {
+        if self.cfg.pipelined_pessimistic_lock {
+            // write to the raft store asynchronously
+            let engine = self.engine.clone();
+            self.spawn(async move {
+                engine.write(txn.into_modifies()).await;
+                drop(guards);
+            });
+        } else {
+            // write to the raft store synchronously
+            self.engine.write(txn.into_modifies()).await; 
+        }
+        ...
+    }
+    resp
+}
+```
+
+The goal is to put the whole process of each command inside a single function. Then, people only need to look at one function to learn the process. Code related to transactions will be more understandable.
+
+Moreover, long code paths and jumps are avoided. It's never a problem that dependencies and configurations need be passed through the long path.
+
+## In-memory lock table
+
+Both the latch and the lock manager stores memory locks and notify when the locks are released. For "parallel commit", we also need another memory locking mechanism. It'll be good to have an integrated locking mechanism handling all these requirements.

```

Comment:
I hesitate a little to get into refactoring to serve parallel commit before we are 100% sure this is the implementation strategy we will use.

Reply:
I think the final plan is exactly to use a memory lock? The only alternative is to save locks to RocksDB and it's unlikely we'll use that approach...

And I think the lock table can be served as the replacement of the latch plus the waiter manager. We can also start.

Reply:
> I think the final plan is exactly to use a memory lock?

For now, yes, but the point in taking an iterative approach is so that we can change the final plan as we go along. Once the first iteration is complete and we benchmark we might decide something different (perhaps it will show that async commit as a whole is not worthwhile, or that the memory lock optimisation would not help since it is not on the hot path, etc.)

> And I think the lock table can be served as the replacement of the latch plus the waiter manager. We can also start.

I agree. I think it is worth doing this work as a refactoring of the existing code. Since it will probably help with async commit, that is a bonus :-)